core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -844,16 +886,31 @@ class GroupCoordinator(val brokerId: Int,
     }
   }
 
-  def tryCompleteHeartbeat(group: GroupMetadata, member: MemberMetadata, heartbeatDeadline: Long, forceComplete: () => Boolean) = {
+  def tryCompleteHeartbeat(group: GroupMetadata, memberId: String, isPending: Boolean, heartbeatDeadline: Long, forceComplete: () => Boolean) = {
     group.inLock {
-      if (member.shouldKeepAlive(heartbeatDeadline) || member.isLeaving)
-        forceComplete()
-      else false
+      if (isPending || !group.has(memberId))
```

Comment:
Could we use `isPendingMember` and get rid of the flag?

Reply:
It's a bit hard. The flag I setup here is to let Heartbeat purgatory be aware whether this is a delayed operation for pending member or for active member. The `pendingMembers` struct could be changed before the delayed operation timeouts, if that happens, `isPendingMember` couldn't know whether this *was* a pending member or not, thus assuming active members timing out and remove member in accident.